### PR TITLE
Default ignores patterns

### PR DESCRIPTION
**Pull request by Shidory**
syncthing default ignores

### CODE REVIEW

File: lib/api/api.go [hunk 1]
```diff
@@ -1173,6 +1174,30 @@ func (s *service) getDBIgnores(w http.ResponseWriter, r *http.Request) {
 	})
 }
 
+func (s *service) getIgnoresFromDeletedFolder(w http.ResponseWriter, r *http.Request) {
```
Comment (calmh) [inline lib/api/api.go:1177]:
```suggestion
func (s *service) getIgnoresRaw(w http.ResponseWriter, r *http.Request) {
```

File: lib/api/api.go [hunk 2]
```diff
@@ -1173,6 +1174,30 @@ func (s *service) getDBIgnores(w http.ResponseWriter, r *http.Request) {
...
+	qs := r.URL.Query()
+
+	path := qs.Get("path")
+	finalPath := path + "/.stignore"
+	if _, err := os.Stat(finalPath); !os.IsNotExist(err) {
```
Comment (calmh) [inline lib/api/api.go:1182]: Don't `stat()` the file beforehand; even if this succeeds, the file may be gone by the time we get to `open()`. Instead just try to open and handle the error by returning from the handler.

File: lib/api/api.go [hunk 3]
```diff
@@ -1173,6 +1174,30 @@ func (s *service) getDBIgnores(w http.ResponseWriter, r *http.Request) {
...
+
+		file, err := os.Open(finalPath)
+		if err != nil {
+			log.Fatal(err)
```
Comment (calmh) [inline lib/api/api.go:1186]: Let's not crash here. :) Instead, something like

```
if err != nil {
  http.Error(w, "Not found", http.StatusNotFound)
  return
}
```

File: lib/api/api.go [hunk 4]
```diff
@@ -1173,6 +1174,30 @@ func (s *service) getDBIgnores(w http.ResponseWriter, r *http.Request) {
...
+		}
+		defer file.Close()
+
+		b, err := ioutil.ReadAll(file)
```
Comment (calmh) [inline lib/api/api.go:1190]: In fact, you can use `ioutil.ReadFile` to get all the file bytes directly without having to open+read+close. We typically call a byte slice `bs`.

File: lib/api/api.go [hunk 5]
```diff
@@ -1173,6 +1174,30 @@ func (s *service) getDBIgnores(w http.ResponseWriter, r *http.Request) {
...
+
+		finalData := string(b[:len(b)])
```
Comment (calmh) [inline lib/api/api.go:1192]: Simpler as just `string(bs)` without the length stuff, and you can do it directly in the map, so:

```
sendJSON(w, map[string]string{
  "ignore": string(bs),
})
```

File: lib/api/api.go [hunk 6]
```diff
@@ -247,6 +247,7 @@ func (s *service) serve(ctx context.Context) {
 	getRestMux.HandleFunc("/rest/db/completion", s.getDBCompletion)              // device folder
 	getRestMux.HandleFunc("/rest/db/file", s.getDBFile)                          // folder file
 	getRestMux.HandleFunc("/rest/db/ignores", s.getDBIgnores)                    // folder
+	getRestMux.HandleFunc("/rest/db/path", s.getIgnoresFromDeletedFolder)        // folder
```
Comment (calmh) [inline lib/api/api.go:250]: This should not be a `/db/` method I think. Maybe `/rest/system/raw-ignore`?

File: gui/default/syncthing/device/editDeviceModalView.html [hunk 7]
```diff
@@ -1,4 +1,4 @@
-<modal id="editDevice" status="default" icon="{{editingExisting ? 'fas fa-pencil-alt' : 'fas fa-desktop'}}" heading="{{editingExisting ? 'Edit Device' : 'Add Device' | translate}} {{currentDevice.name}}" large="yes" closeable="yes">
+<modal id="editDevice" status="default" icon="{{editingExisting ? 'fas fa-pencil-alt' : 'fas fa-desktop'}}" heading="{{editingExisting ? 'Edit Device : ' : 'Add Device' | translate}} {{currentDevice.name}}" large="yes" closeable="yes">
```
Comment (calmh) [inline gui/default/syncthing/device/editDeviceModalView.html:1]: Looks like a change from another branch? remove

File: gui/default/syncthing/core/syncthingController.js [hunk 8]
```diff
@@ -1789,13 +1791,30 @@ angular.module('syncthing.core')
         };
 
         $scope.addFolder = function () {
+            $('#folderLabel').change(function() {
+                var path = $('#folderPath').val();
+                $http.get(urlbase + '/db/path?path=' + encodeURIComponent(path))
+                .success(function (data) {
+                    if(data.ignore){
+                        $('#folder-ignores textarea').val(data.ignore);
```
Comment (calmh) [inline gui/default/syncthing/core/syncthingController.js:1799]: Remove the disabled attr here as well? We've loaded the ignore patterns, so we might as well let the user edit them before creating the folder

File: gui/default/syncthing/core/syncthingController.js [hunk 9]
```diff
@@ -1768,6 +1769,7 @@ angular.module('syncthing.core')
                 })
                 .error(function (err) {
                     $('#folder-ignores textarea').val($translate.instant("Failed to load ignore patterns."));
+                    $('#folder-ignores textarea').removeAttr('disabled');
```
Comment (calmh) [inline gui/default/syncthing/core/syncthingController.js:1772]: Probably should not make this change. It's part of the other issue (editing ignore patterns that fail to load...)

Comment (Shidory) [reply gui/default/syncthing/core/syncthingController.js:1772]: We've also already worked on the editing ignore patterns that fail to load

Comment (calmh) [reply gui/default/syncthing/core/syncthingController.js:1772]: I know, but lets make that a separate change & PR

File: gui/default/syncthing/core/syncthingController.js (hunk 8 above)
Comment (Shidory) [reply gui/default/syncthing/core/syncthingController.js:1799]: Already at this level the user can modify the ignores.

File: gui/default/syncthing/core/syncthingController.js (hunk 9 above)
Comment (Shidory) [reply gui/default/syncthing/core/syncthingController.js:1772]: On the same branch or on another ?

File: lib/api/api.go (hunk 2 above)
Comment (Shidory) [reply lib/api/api.go:1182]: can you clarify what you mean by that?

File: gui/default/syncthing/core/syncthingController.js (hunk 9 above)
Comment (calmh) [reply gui/default/syncthing/core/syncthingController.js:1772]: Separate issues => separate branches, separate PRs

File: gui/default/syncthing/core/syncthingController.js [hunk 10]
```diff
@@ -1705,6 +1705,7 @@ angular.module('syncthing.core')
         };
 
         $scope.editFolder = function (folderCfg) {
+            console.log($scope.config);
```
Comment (calmh) [inline gui/default/syncthing/core/syncthingController.js:1708]: remove

File: lib/api/api.go [hunk 11]
```diff
@@ -1173,6 +1173,7 @@ func (s *service) getDBIgnores(w http.ResponseWriter, r *http.Request) {
 	})
 }
 
+
```
Comment (calmh) [inline lib/api/api.go:1176]: remove

File: gui/default/syncthing/device/editDeviceModalView.html (hunk 7 above)
Comment (calmh) [reply gui/default/syncthing/device/editDeviceModalView.html:1]: not addressed

File: lib/api/api.go (hunk 2 above)
Comment (calmh) [reply lib/api/api.go:1182]: You can remove the whole outer `if ... os.Stat()` thing. Either the file exists, and then you don't need it, or it does exist, and then you get the same error from `ioutil.ReadFile`. If you want to "mask" the does-not-exist error as a successful empty return, then do the `os.IsNotExist()` check on *that* error.

File: gui/default/syncthing/core/syncthingController.js (hunk 8 above)
Comment (calmh) [reply gui/default/syncthing/core/syncthingController.js:1797]: But we set the `disabled` attributes sometimes (when there is an error loading the folder), and we reuse the same modal for adding a folder here, so I think we still need to remove the `disabled` attribute, even if we think it's probably not set.

File: lib/api/api.go [hunk 12]
```diff
@@ -1173,6 +1174,21 @@ func (s *service) getDBIgnores(w http.ResponseWriter, r *http.Request) {
 	})
 }
 
+func (s *service) getIgnoresRaw(w http.ResponseWriter, r *http.Request) {
+	qs := r.URL.Query()
+
+	path := qs.Get("raw-ignore")
+	finalPath := path + "/.stignore"
+	bs, err := ioutil.ReadFile(finalPath)
```
Comment (calmh) [inline lib/api/api.go:1182]: Let's call the query param `path` and use filepath.Join:

```suggestion
	folderPath := qs.Get("path")
	ignoresPath := filepath.Join(folderPath, ".stignore")
	bs, err := ioutil.ReadFile(ignoresPath)
```

File: lib/api/api.go [hunk 13]
```diff
@@ -247,6 +247,7 @@ func (s *service) serve(ctx context.Context) {
 	getRestMux.HandleFunc("/rest/db/completion", s.getDBCompletion)              // device folder
 	getRestMux.HandleFunc("/rest/db/file", s.getDBFile)                          // folder file
 	getRestMux.HandleFunc("/rest/db/ignores", s.getDBIgnores)                    // folder
+	getRestMux.HandleFunc("/rest/system/raw-ignore", s.getIgnoresRaw) 		     // folder
```
Comment (calmh) [inline lib/api/api.go:250]:
```suggestion
	getRestMux.HandleFunc("/rest/db/ignores/raw", s.getIgnoresRaw) 		     // path
```

Comment (calmh) [reply lib/api/api.go:250]: The ignores strictly speaking don't live in the `db`, but they probably should and we can keep the API endpoint there for the time being... sorry for changing my mind on this.

**Comment by AudriusButkevicius** [comment]
Can someone give a brief intro how this is supposed to work?

**Comment by Shidory** [comment]
We are supposed to give the user the possibility to create a default ignore pattenrs that would apply when creating a new folder

**Comment by Shidory** [comment]
And if the folder had already been created and deleted, give the possibility to retrieve the ignorance that was applied to it.

File: lib/api/api.go [hunk 14]
```diff
@@ -1173,6 +1174,21 @@ func (s *service) getDBIgnores(w http.ResponseWriter, r *http.Request) {
 	})
 }
 
+func (s *service) getIgnoresRaw(w http.ResponseWriter, r *http.Request) {
+	qs := r.URL.Query()
+
+	folderPath := qs.Get("raw-ignore")
```
Comment (calmh) [inline lib/api/api.go:1180]:
```suggestion
	folderPath := qs.Get("path")
```

Again, lets call this `path`, please.

**Comment by calmh** [comment]
@Shidory just needs a gofmt to pass that check as well.

@AudriusButkevicius There is a new field for default ignore patterns in the global settings. When creating a new folder we first try to load an existing .stignore (to avoid clobbering) and otherwise fill with the default ignore patterns. Which are empty by default, which matches the current behavior.

**Comment by AudriusButkevicius** [comment]
Wouldn't it be better to do it in go?

Now anyone using the API/cli/rest will have to mimic the same behaviour.

**Comment by calmh** [comment]
Do what in go? We already set all folder options including initial ignore patterns via the api (GUI).

**Comment by AudriusButkevicius** [comment]
Ah yes, we no longer have any "defaults" in Go, so I guess this approach makes sense, even if it's sad.

**Comment by calmh** [comment]
Well, I mean, technically it would be totally possible to apply the defaults from the config if the GUI sends a folder create with an empty set of ignore patterns, or maybe a magic flag. But I think that would be a much more surprising behavior and probably trickier to implement, as the physical folder creation is far separated from the config POST. And raises questions like should the patterns be applied when a new folder is just added to the config and syncthing restarted? That would seem to be better served then by a folder flag "use default ignores" and have them be always applied from the config instead of the .stignore. Would that be better?

**Comment by AudriusButkevicius** [comment]
Not sure, it's probably good as it is, but I think all of the logic for folder creation/defaults being in the UI is a bad state to be in.

I know it's already an existing state, so I was just questioning whether we should keep piling on top of that.

In an ideal world you'd call `folder/defaults` which gives you the prepopulated default values, you tweak them in the UI and post them back I guess.

**Comment by imsodin** [comment]
there is an issue about configurable folder defaults. i think a first step towards that would be exposing something like /rest/folder/default/config and then we could also have  /rest/folder/default/ignores. seems unrelated to this pr though

**Comment by calmh** [comment]
Right. So in the interest of keeping things sane, we postpone this issue until a system for actual folder defaults exist, at which point we can revisit how to handle the default ignore patterns specifically.

**Pull request closed without merging**